test/replication-luatest/gh_6754_promote_before_new_term_test.lua

```diff
@@ -0,0 +1,59 @@
+local luatest = require('luatest')
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+local g = luatest.group('gh-6754-promote-before-new-term')
+
+g.before_all(function(g)
+    g.cluster = cluster:new({})
+
+    g.box_cfg = {
+        read_only = true,
+        election_mode = 'off',
+        replication_synchro_quorum = 1,
+        replication_timeout = 0.1,
+        replication = {
+            server.build_instance_uri('server_1'),
+            server.build_instance_uri('server_2'),
+        },
+    }
+
+    g.server_1 = g.cluster:build_and_add_server(
+        {alias = 'server_1', box_cfg = g.box_cfg})
+
+    g.box_cfg.read_only = false
+
+    g.server_2 = g.cluster:build_and_add_server(
+        {alias = 'server_2', box_cfg = g.box_cfg})
+    g.cluster:start()
+    g.cluster:wait_fullmesh()
+end)
+
+g.after_all(function(g)
+    g.cluster:stop()
+end)
+
+g.test_promote_new_term_order = function(g)
+    local lsn = g.server_2:lsn()
+    local f = g.server_1:exec(function()
+        box.error.injection.set('ERRINJ_RELAY_FROM_TX_DELAY', true)
+        local f = require('fiber').new(box.ctl.promote)
```

Comment:
I don't see where `promote` would hang. 
ERRINJ_RELAY_FROM_TX_DELAY simply blocks relay from pushing election term, but doesn't block the node's promote().

I suppose these are leftovers from the previous patch version, where promote() would block until the term is broadcast.

I applied the following changes, and the test still passed:
```diff
diff --git a/test/replication-luatest/gh_6754_promote_before_new_term_test.lua b/test/replication-luatest/gh_6754_promote_before_new_term_test.lua
index 20682dbe8..d2f6d98b6 100644
--- a/test/replication-luatest/gh_6754_promote_before_new_term_test.lua
+++ b/test/replication-luatest/gh_6754_promote_before_new_term_test.lua
@@ -34,24 +34,21 @@ end)
 
 g.test_promote_new_term_order = function(g)
     local lsn = g.server_2:lsn()
-    local f = g.server_1:exec(function()
+    g.server_1:exec(function()
         box.error.injection.set('ERRINJ_RELAY_FROM_TX_DELAY', true)
-        local f = require('fiber').new(box.ctl.promote)
-        f:set_joinable(true)
+        box.ctl.promote()
         box.cfg{read_only = false}
         box.schema.create_space('test'):create_index('pk')
         box.space.test:replace({1})
-        return f:id()
     end)
     g.server_2:wait_lsn(g.server_1, lsn + 2)
 
     local election_term = g.server_2:election_term()
     local synchro_queue_term = g.server_2:synchro_queue_term()
 
-    g.server_1:exec(function(f)
+    g.server_1:exec(function()
         box.error.injection.set('ERRINJ_RELAY_FROM_TX_DELAY', false)
-        require('fiber').find(f):join();
-    end, {f})
+    end)
     g.server_2:wait_election_term(synchro_queue_term)
 
     luatest.assert(synchro_queue_term <= election_term,
```

Reply:
Thanks for review!
Calling promote in new fiber is indeed for the case if box.ctl.promote() blocks (this test works for #6858 too). I would leave it as is, because this way test doesn't rely on box.ctl.promote() not blocking.

Reply:
I think it's better to stumble upon this blocking when (and if) promote behaviour changes. This way we'll be able to fix the test with understanding why promote() became blocking.

On the contrary, if you leave the current version, the test might continue passing even if promote blocking is introduced by some error.

Let's wait for @Gerold103's opinion.

Reply:
I agree with Sergey here. Promotion in a fiber will hide potential issues if something relevant changes in promotion behaviour. I propose to drop the fiber if the test is useful without it.